### ISSUE TITLE
v1.0.0

### ISSUE THREAD

**Issue by jscottsmith** [jscottsmith/react-observed#1]
@TODO for `v1.0.0` release:

- [x] Publish a demo and/or Storybook
- [x] Add test coverage by [codecov.io](https://codecov.io/)
- [x] Support React 16
- [x] Test and document browser support
- [x] Proofread and clarify prop behavior in README (See #2, #3)
- [x] Add a story to demo `once` prop
- [x] Merge incoming `props.options` with defaults
- [x] Change defaults: ~~options.threshold to 0 and~~ `intersectionRatio` to `0` (See #2)
- [x] Default `options.threshold` to `intersectionRatio` when none is provided
- [x] add handlers for `isInView` state changes:
  - `onEnter`
  - `onExit`
  - `onChange`
- [x] add handler for `onIntersect` that's called with [entry](https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserverEntry) data when each threshold is met

**Comment by oyeanuj** [comment]
@jscottsmith I was wondering if you are thinking of adding a callback for when an element gets in view? Like an `onObserve` maybe?

I see myself needing that for 'load more/infinite loaders' like situations.

**Comment by jscottsmith** [comment]
Ya I think a callback prop function would make such use cases simpler. Not sure about name -- `onObserve` is probably not be descriptive enough. Other options:

- `onInViewUpdate`
- `onInViewChange`
- `onIsInViewUpdate`
- `onObservedUpdate`

Leaning toward something with `inView` in the name to mirror the render callback boolean name, `isInView`.

**Comment by oyeanuj** [comment]
@jscottsmith Totally, I just threw out the first name that came to mind. But on thinking more, you could go for `onChange` which passes a boolean param like `inView` to the user, or `onViewEnter`, `onViewExit` - more wordy but may match more patterns in other libraries?

**Comment by jscottsmith** [comment]
Good ideas, I like `onViewEnter`/`onViewExit`/`onViewChange`. Kinda treats the observer like a waypoint utility.

**Comment by jscottsmith** [comment]
@oyeanuj published a new beta with the handlers and updated observer defaults, if you wanna try it out.

**Comment by oyeanuj** [comment]
@jscottsmith thank you! it seems to work well, maybe the only thing which needs clarification is for users might be how to think about `intersectionRatio` - how it ends up being different than `threshold` or in what cases it would be different than `threshold`?  Oh, and should it default to `threshold` value if not passed?

**Comment by jscottsmith** [comment]
Ya true, that might be clearer. 

What I could do is make the `threshold` equal to the `intersectionRatio ` (probably not the other way around since it won't work with an array) if the `threshold` isn't explicitly passed as a prop `option`. 

They could always be the same but I'd like to keep the possibility of having the `threshold` be an array for the sake of flexibility -- perhaps having a handler for each intersection entry even when it doesn't meet the ratio, or passing the current intersection ratio in the render callback.

So do you think that will help clarify/simplify usage?

To be clear, this is what it would look like if you only had to set the ratio.

```jsx
<Observed intersectionRatio={0.5}>
    {({ mapRef }) => (
        <div ref={mapRef} />
    )}
</Observed>

// options.threshold will equal 0.5

```
and if you needed to overwrite the threshold behavior for whatever reason you could just pass the prop `options.threshold`:

```jsx
<Observed intersectionRatio={0.5} options={{
    threshold: [0, 0.25, 0.5, 0.75, 1], 
}}>
    {({ mapRef }) => (
        <div ref={mapRef} />
    )}
</Observed>
```

I think that could work -- it's certainly simpler but also doesn't limit future usage.

**Comment by oyeanuj** [comment]
Yes, I think this looks good! I think something to clarify would be in the second case, what should the user expect to happen when 25% of the observed object is in view - will the callbacks fire but `isInView` be `false`? Essentially, what happens when `intersectionRatio` is greater than the threshold? (or even vice versa)?

Also, I noticed you added the fix for the `disconnect` error - is it possible to push a release to npm for that? Thank you!

**Comment by jscottsmith** [comment]
> what happens when intersectionRatio is greater than the threshold? (or even vice versa)

Nothing. I originally thought that it was necessary to provide additional thresholds to improve accuracy because I thought there was the potential that they might not all be called. However that impression was just due to a bug in my code 😄. Also, all the examples I saw on MDN used an array of thresholds but as far as I can tell only a single value is needed for most use cases.

However I do think it still may be useful to have an array of thresholds. Perhaps to have an `onIntersect` handler that is called when each threshold is met. Maybe even pass the current intersect value into the render callback.

>  I noticed you added the fix for the disconnect error - is it possible to push a release to npm for that?

Ya of course, I'll publish shortly.

**Comment by jscottsmith** [comment]
Published [v1.0.0](https://github.com/jscottsmith/react-observed/releases/tag/v1.0.0)
